cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/Operand.qll

```diff
@@ -406,6 +412,11 @@ class ArgumentOperand extends RegisterOperand {
  */
 class ThisArgumentOperand extends ArgumentOperand {
   override ThisArgumentOperandTag tag;
+
+  // in some cases the def location seems to make more sense, but most of the
+  // time it does not really make a difference, and on some occations the def
+  // has no location at all, so that in general it is better use the use location
+  override Language::Location getLocation() { result = this.getUse().getLocation() }
```

Comment:
Yeah, I guess there won't be a location if there isn't an explicit `this` parameter?

Would it make sense to use the location of the definition if it exists (and isn't an `UnknownLocation`), and otherwise pick the location of the use? Hopefully, we then get the location of an explicit `this` qualifier if it exists, and otherwise, we get the location of the `call` if there is no explicit qualifier?

Reply:
from what I've seen, it's true that in typical method calls with an explicit instance the definition location was indeed sensible, for example
```
this->foo()
~~~~
instance.do_something()
~~~~~~~~
```
this got a bit worse in method calls with an implicit instance (for example, method calls in the same class namespace), and even worse in implicit method calls (for example, constructors and operators). Although UnknownLocations where the worst instances, there were other that though not unknown, were also not really helpful.

I'll try to dig in on this and see if I can come up with a smarter getLocation implementation for this case.

Reply:
Sounds good. We don't need to put too much smarts into the `getLocation` predicate, but it would be nice to have the simple cases of explicit and implicit qualifiers working (by using `getDef()` when the location makes intuitive sense, and otherwise falling back to `getUse` like we currently do).

Reply:
Going back to analyzing the differences between `getUse` and `getAnyDef` locations for `ThisArgumentOperand`, turns out that apart from the `UknownLocation` cases for `getAnyDef`, the latter is generally a better fit for the location. As far as I can see, it will result in an `UknonwnLocation` for implicit calls to base class constructors/operators in default definitions of the same in the derived class. That is for example:
```
class Base {};
class Derived : public Base {};
```
the default constructors of `Derived` will have a call to the corresponding constructors of `Base`, and those calls will have `getAnyDef` with an `UnknownLocation`, where `getUse` will point instead to the beginning of `Derived`'s definition.

So your guess of using `getAnyDef` and defaulting to `getUse` on `UknownLocation` works really good, I'll implement it now.

Reply:
That sounds great!

Reply:
There is also this weird corner case from `ir.cpp` that I don't fully understand:
```c++
void temporary_hierarchy() {
    POD_Base b = returnValue<POD_Middle>();
    b = (returnValue<POD_Derived>());  // Multiple conversions plus parens
    int x = returnValue<POD_Derived>().x;
    float f = (returnValue<POD_Derived>()).f();
}
``` 
The last line, `getAnyDef()` of the `ThisArgumentOperand` for the call to `f` does not have a location :confused: Probably because it's a temporary object. Still I think it will be good enough (`getUse` points at `f`, so still in the general vicinity).

Reply:
Yeah, I've also seen missing locations from temporary objects before. They do have location information _sometimes_, and I haven't found a consistent pattern in when they're missing. Feel free to open a new issue with a list of places where the lack of a location is a problem. I feel like that could potentially be a `good first issue` in the extractor for new starters.